### PR TITLE
Handle symbol reference operations in CFG

### PR DESCRIPTION
**Pull request by mavasani**
`ILocalReferenceOperation`, `IParameterReferenceOperation`, `IFieldReferenceOperation`, `IPropertyReferenceOperation`, `IMethodReferenceOperation` and `IEventReferenceOperation`

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventReferenceOperation.cs [hunk 1]
```diff
@@ -201,5 +201,153 @@ public static void M()
             VerifyOperationTreeAndDiagnosticsForTest<MemberAccessExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
         }
 
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventReferenceOperation.cs]: What about adding or removing delegate from events? Or is that going to be a separate PR?

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventReferenceOperation.cs]: Yes, I plan to handle `IEventAssignmentOperation` next.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -4022,17 +4022,23 @@ public override IOperation VisitCaughtException(ICaughtExceptionOperation operat
...
+            (IOperation visitedInstance, ImmutableArray<IArgumentOperation> visitedArguments) = VisitInstanceWithArguments(operation.Instance, operation.Arguments);
+            return new InvocationExpression(operation.TargetMethod, visitedInstance, operation.IsVirtual, visitedArguments, semanticModel: null, operation.Syntax,
+                                            operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        private (IOperation, ImmutableArray<IArgumentOperation>) VisitInstanceWithArguments(IOperation instance, ImmutableArray<IArgumentOperation> arguments)
+        {
+            if (instance != null)
             {
                 // PROTOTYPE(dataflow): drop instance for a static method
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >// PROTOTYPE(dataflow): drop instance for a static method [](start = 16, length = 57)

Please implement this work item and add relevant tests

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Done - I have pushed a new commit addressing the prototype comments and adding relevant unit tests for all affected operation nodes.

---
In reply to: [187757013](https://github.com/dotnet/roslyn/pull/26797#discussion_r187757013) [](ancestors = 187757013)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests.cs [hunk 3]
```diff
@@ -420,5 +420,248 @@ void M(A a)
             // Verify we return null operation for child nodes of member access expression.
             Assert.Null(model.GetOperation(expr.Name));
         }
+
+        [CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)]
+        [Fact]
+        public void MethodReference_NoControlFlow()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests.cs]: >MethodReference_NoControlFlow [](start = 20, length = 29)

Consider creating a file dedicated to method references and moving the tests there.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests.vb [hunk 4]
```diff
@@ -909,5 +909,230 @@ Option Strict On 'BIND:"Option Strict On"
 
             VerifyNoOperationTreeForTest(Of StatementSyntax)(source)
         End Sub
+
+        <CompilerTrait(CompilerFeature.IOperation, CompilerFeature.Dataflow)>
+        <Fact>
+        Public Sub MethodReference_NoControlFlow()
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests.vb]: >MethodReference_NoControlFlow [](start = 19, length = 29)

Same comment as for C#

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPropertyReferenceExpression.cs [hunk 5]
```diff
@@ -350,5 +350,458 @@ public static void Main()
...
+        [Fact]
+        public void PropertyReference_ControlFlowInReceiverAndArguments()
+        {
+            string source = @"
+class C
+{
+    public int this[int i1, int i2] => 0;
+    void M(C c1, C c2, int? i1, int i2, int? i3, int i4, int p)
+    /*<bind>*/{
+        p = (c1 ?? c2)[i1 ?? i2, i3 ?? i4];
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPropertyReferenceExpression.cs]: >i1 ?? i2, i3 ?? i4 [](start = 23, length = 18)

it would be good to have a test where there is a control flow in the second argument and no control flow in receiver and the first argument

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPropertyReferenceExpression.cs]: Sure, will add one.

---
In reply to: [188124271](https://github.com/dotnet/roslyn/pull/26797#discussion_r188124271) [](ancestors = 188124271)

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPropertyReferenceExpression.cs]: Added `PropertyReference_ControlFlowInSecondArgument`

---
In reply to: [188126611](https://github.com/dotnet/roslyn/pull/26797#discussion_r188126611) [](ancestors = 188126611,188124271)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 6]
```diff
@@ -4022,17 +4022,23 @@ public override IOperation VisitCaughtException(ICaughtExceptionOperation operat
...
         public override IOperation VisitInvocation(IInvocationOperation operation, int? captureIdForResult)
         {
-            if (operation.Instance != null)
+            IOperation instance = operation.TargetMethod.IsStatic ? null : operation.Instance;
+            (IOperation visitedInstance, ImmutableArray<IArgumentOperation> visitedArguments) = VisitInstanceWithArguments(instance, operation.Arguments);
+            return new InvocationExpression(operation.TargetMethod, visitedInstance, operation.IsVirtual, visitedArguments, semanticModel: null, operation.Syntax,
+                                            operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        private (IOperation, ImmutableArray<IArgumentOperation>) VisitInstanceWithArguments(IOperation instance, ImmutableArray<IArgumentOperation> arguments)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider adding names to the tuple elements. My personal rule of thumb (not sure we have an established policy here). If it's on a non-local function then generally best to add names.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 7]
```diff
@@ -4022,17 +4022,23 @@ public override IOperation VisitCaughtException(ICaughtExceptionOperation operat
 
         public override IOperation VisitInvocation(IInvocationOperation operation, int? captureIdForResult)
         {
-            if (operation.Instance != null)
+            IOperation instance = operation.TargetMethod.IsStatic ? null : operation.Instance;
+            (IOperation visitedInstance, ImmutableArray<IArgumentOperation> visitedArguments) = VisitInstanceWithArguments(instance, operation.Arguments);
+            return new InvocationExpression(operation.TargetMethod, visitedInstance, operation.IsVirtual, visitedArguments, semanticModel: null, operation.Syntax,
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >semanticModel [](start = 124, length = 13)

Why do we not add a semantic model here?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventReferenceOperation.cs [hunk 8]
```diff
@@ -201,5 +201,217 @@ public static void M()
...
+    Jump if True (Regular) to Block[B3]
+        IIsNullOperation (OperationKind.IsNull, Type: System.Boolean, IsImplicit) (Syntax: 'c1')
+          Operand: 
+            IFlowCaptureReferenceOperation: 1 (OperationKind.FlowCaptureReference, Type: C, IsImplicit) (Syntax: 'c1')
+
+    Next (Regular) Block[B2]
+Block[B2] - Block
+    Predecessors: [B1]
+    Statements (1)
+        IFlowCaptureOperation: 2 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'c1')
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventReferenceOperation.cs]: This looks like a second capture of `c1` here. Is that intended? If so can you elaborate why here? At a glance I would expect to see a reference to the previous flow capture.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 7 above)
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > Why do we not add a semantic model here? 

We intentionally remove SemanticModel from all nodes. Right now it is used only to lazily calculate Parent for a node based on original IOperation tree and syntax, this won't work with rewritten tree and shouldn't even be attempted. We are using SemanticModel in this rewriter, but it won't be possible to feed any of the rewritten nodes back into this rewriter.  

---
In reply to: [188160319](https://github.com/dotnet/roslyn/pull/26797#discussion_r188160319) [](ancestors = 188160319)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 9]
```diff
@@ -4538,7 +4544,48 @@ public override IOperation VisitLiteral(ILiteralOperation operation, int? captur
...
+        public override IOperation VisitMethodReference(IMethodReferenceOperation operation, int? captureIdForResult)
+        {
+            IOperation visitedInstance = operation.Method.IsStatic ? null : Visit(operation.Instance);
+            return new MethodReferenceExpression(operation.Method, operation.IsVirtual, visitedInstance, semanticModel: null,
+                                                 operation.Syntax, operation.Type, operation.ConstantValue, IsImplicit(operation));
+        }
+
+        public override IOperation VisitPropertyReference(IPropertyReferenceOperation operation, int? captureIdForResult)
+        {
+            IOperation instance = operation.Property.IsStatic ? null : operation.Instance;
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider just modifying VisitInstanceWithArguments to take an ISymbol, and doing the check there, instead of duplicating here and in VisitInvocation. You could also consider having it accept an empty or default array of arguments, and reusing it for visiting all instances, so we can have the static check in one place for all.

Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I considered that initially, but given the fact that the static check is pretty trivial for callers, `VisitInstanceWithArguments` is currently used only at 2 places, and it has a much cleaner API to implementation mapping, I chose the current structure. Let me know if you feel strongly about your suggestion and I can refactor in my next commit (if there is additional feedback) or next PR.

---
In reply to: [188427893](https://github.com/dotnet/roslyn/pull/26797#discussion_r188427893) [](ancestors = 188427893)

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Well, I also meant to consider using VisitInstanceWithArguments in more places as well, ie make it work for VisitEventReference, VisitMethodReference, etc. I do think it would be good to have one location where these checks are done so it's easier to verify, a follow up is fine.

---
In reply to: [188429842](https://github.com/dotnet/roslyn/pull/26797#discussion_r188429842) [](ancestors = 188429842,188427893)

**Review by 333fred — APPROVED** [review]
:shipit:

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 9 above)
Comment (mavasani) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > I also meant to consider using VisitInstanceWithArguments in more places as well, ie make it work for VisitEventReference, VisitMethodReference, etc.

That sounds a bit convoluted to me, especially since they can never have arguments. I think your approach would be more valuable if the shared code grows more than one line: `IOperation visitedInstance = operation.Event.IsStatic ? null : Visit(operation.Instance);`

---
In reply to: [188431334](https://github.com/dotnet/roslyn/pull/26797#discussion_r188431334) [](ancestors = 188431334,188429842,188427893)

**Comment by mavasani** [comment]
@jaredpar Any further concerns?

**Comment by mavasani** [comment]
I am going to merge this in and address any further feedback as a follow-up change.
